astra/src/main/java/com/slack/astra/metadata/core/AstraMetadataStore.java

```diff
@@ -1,331 +1,708 @@
-      cachedModeledFramework = modeledClient.cached();
-      cachedModeledFramework.listenable().addListener(initializedListener, cacheInitializedService);
-      cachedModeledFramework.start();
+    this.zkStore = zkStore;
+    this.etcdStore = etcdStore;
+
+    // Override mode if one of the stores is null
+    if (zkStore == null && etcdStore != null) {
+      this.mode = MetadataStoreMode.ETCD_EXCLUSIVE;
+      LOG.info("ZK store is null, overriding mode to ETCD_EXCLUSIVE regardless of configured mode");
+    } else if (etcdStore == null && zkStore != null) {
+      this.mode = MetadataStoreMode.ZOOKEEPER_EXCLUSIVE;
+      LOG.info(
+          "Etcd store is null, overriding mode to ZOOKEEPER_EXCLUSIVE regardless of configured mode");
+    } else if (etcdStore == null && zkStore == null) {
+      throw new IllegalArgumentException("Both zkStore and etcdStore cannot be null");
     } else {
-      cachedModeledFramework = null;
-      cacheInitializedService = null;
+      this.mode = mode;
     }
+
+    this.meterRegistry = meterRegistry;
   }
 
+  /**
+   * Creates a new metadata node.
+   *
+   * @param metadataNode the node to create
+   * @return a CompletionStage that completes when the operation is done
+   */
   public CompletionStage<String> createAsync(T metadataNode) {
-    // by passing the version 0, this will throw if we attempt to create and it already exists
-    return modeledClient.set(metadataNode, 0);
+    switch (mode) {
+      case ZOOKEEPER_EXCLUSIVE:
+        return zkStore.createAsync(metadataNode);
+      case ETCD_EXCLUSIVE:
+        return etcdStore.createAsync(metadataNode);
+      case BOTH_READ_ZOOKEEPER_WRITE:
+        // In migration to ZK mode, writes only go to ZK
+        return zkStore.createAsync(metadataNode);
+      case BOTH_READ_ETCD_WRITE:
+        // In migration to Etcd mode, writes only go to Etcd
+        return etcdStore.createAsync(metadataNode);
+      default:
+        throw new IllegalArgumentException("Unknown metadata store mode: " + mode);
+    }
   }
 
+  /**
+   * Synchronously creates a new metadata node.
+   *
+   * @param metadataNode the node to create
+   */
   public void createSync(T metadataNode) {
-    try {
-      this.createCall.increment();
-
-      createAsync(metadataNode)
-          .toCompletableFuture()
-          .get(zkConfig.getZkConnectionTimeoutMs(), TimeUnit.MILLISECONDS);
-    } catch (InterruptedException | ExecutionException | TimeoutException e) {
-      throw new InternalMetadataStoreException("Error creating node " + metadataNode, e);
+    switch (mode) {
+      case ZOOKEEPER_EXCLUSIVE:
+        zkStore.createSync(metadataNode);
+        break;
+      case ETCD_EXCLUSIVE:
+        etcdStore.createSync(metadataNode);
+        break;
+      case BOTH_READ_ZOOKEEPER_WRITE:
+        // In migration to ZK mode, writes only go to ZK
+        zkStore.createSync(metadataNode);
+        break;
```

Comment:
Do we need a mode where it writes to both? Otherwise we go directly from only writing to ZK (reading from both except we won't have data in ETCD) to only writing to ETCD

Reply:
Writing to both gets to be extremely complicated to manage and reason about state - especially differences in how ephemeral nodes work. My current plan is a point-in-time where everything _new_ is cut over, and then we'll have a migration step to force move some of the stores that rarely change (ie, dataset).